invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java

```diff
@@ -199,75 +201,78 @@ protected UsageInArrearAggregate getToBeBilledUsageDetails(final LocalDate start
             }
-            if (tierNum == 1 || nbUsedTierBlocks > 0) {
+            if (tierNum == 1 || nbUsedTierBlocks.compareTo(BigDecimal.ZERO) > 0) {
                 toBeBilledDetails.add(new UsageConsumableInArrearTierUnitAggregate(tierNum, tieredBlock.getUnit().getName(), tieredBlock.getPrice().getPrice(getCurrency()), blockTierSize, nbUsedTierBlocks));
             }
         }
         return toBeBilledDetails;
     }
 
-    UsageConsumableInArrearTierUnitAggregate computeToBeBilledConsumableInArrearWith_TOP_TIER(final List<TieredBlock> tieredBlocks, final List<UsageConsumableInArrearTierUnitAggregate> previousUsage, final Long units) throws CatalogApiException {
-
-        Long remainingUnits = units;
+    UsageConsumableInArrearTierUnitAggregate computeToBeBilledConsumableInArrearWith_TOP_TIER(final List<TieredBlock> tieredBlocks, final List<UsageConsumableInArrearTierUnitAggregate> previousUsage, final BigDecimal units) throws CatalogApiException {
+        BigDecimal remainingUnits = units;
         // By default last last tierBlock
         TieredBlock targetBlock = tieredBlocks.get(tieredBlocks.size() - 1);
         int targetTierNum = tieredBlocks.size();
         int tierNum = 0;
         // Loop through all tier block
         for (final TieredBlock tieredBlock : tieredBlocks) {
-
             tierNum++;
-            final long blockTierSize = tieredBlock.getSize().longValue();
-            final long tmp = remainingUnits / blockTierSize + (remainingUnits % blockTierSize == 0 ? 0 : 1);
-            if ( tmp > tieredBlock.getMax()) { /* Includes the case where max is unlimited (-1) */
-                remainingUnits -= tieredBlock.getMax().longValue() * blockTierSize;
+            final BigDecimal blockTierMax = tieredBlock.getMax();
+            final BigDecimal blockTierSize = tieredBlock.getSize();
+            final BigDecimal[] divRemaining =  remainingUnits.divideAndRemainder(blockTierSize);
+            final BigDecimal tmp =  (divRemaining[1].compareTo(BigDecimal.ZERO) == 0) ? divRemaining[0] : divRemaining[0].add(BigDecimal.ONE);
+            if (tmp.compareTo(blockTierMax) > 0) { /* Includes the case where max is unlimited (-1) */
+                remainingUnits = remainingUnits.subtract(blockTierMax.multiply(blockTierSize));
             } else {
                 targetBlock = tieredBlock;
                 targetTierNum = tierNum;
                 break;
             }
         }
         final long lastBlockTierSize = targetBlock.getSize().longValue();
-        final long nbBlocks = units / lastBlockTierSize + (units % lastBlockTierSize == 0 ? 0 : 1);
-        return new UsageConsumableInArrearTierUnitAggregate(targetTierNum, targetBlock.getUnit().getName(), targetBlock.getPrice().getPrice(getCurrency()), targetBlock.getSize().longValue(), nbBlocks);
+        final long nbBlocks = units.longValue() / lastBlockTierSize + (units.longValue() % lastBlockTierSize == 0 ? 0 : 1);
```

Comment:
@sbrossie do you think calculation here should be converted to BigDecimal as well? 

If so, what happened if this line
```java
units.longValue() / lastBlockTierSize // which will be converted to: units.divide(lastBlockTierSize).add( ..... );
```

Having fractional number left? For example, [this test line](https://github.com/killbill/killbill/blob/d1c910ba7828e0baa8fc79dfbad1b46ae9591dec/invoice/src/test/java/org/killbill/billing/invoice/usage/TestContiguousIntervalConsumableInArrear.java#L277) will return `51.001` instead of `51.0`.

Reply:
@xsalefter Yes, thank you for pointing that out. This should also be changed - all the computation should be done using `BigDecimal` (no more `longValue()` or `doubleValue()` conversion)


So, I think this is as simple as replacing the 2 lines:

```
        final long lastBlockTierSize = targetBlock.getSize().longValue();
        final long nbBlocks = units.longValue() / lastBlockTierSize + (units.longValue() % lastBlockTierSize == 0 ? 0 : 1);

```

With
```
        final BigDecimal lastBlockTierSize = targetBlock.getSize();
        final BigDecimal[] divRemaining =  units.divideAndRemainder(lastBlockTierSize);
        final BigDecimal nbBlocks =  (divRemaining[1].compareTo(BigDecimal.ZERO) == 0) ? divRemaining[0] : divRemaining[0].add(BigDecimal.ONE);
```
(And removing the now unecessary `BigDecimal.valueOf` conversion below)